### PR TITLE
ECS exec: UX bug fixes 

### PR DESCRIPTION
**Pull request by gsweetwood**
## Problem
* #2374 
* Commas injected into output from command execution
* Context menu allows command execution when parent service is not enabled for ECS Exec
* No notification that command is running 
* Warning message is a modal and in-between steps

## Solution
* Implemented wizard for ECS Exec flow. This addresses #2374 (no longer allows for empty task ID) and including the warning message.
* New contexts were created that change depending on if an ECS Service is enabled or disabled for ECS Exec. The option to run a command in a container is only available if the Service is enabled.
* Datestamp and details added to output channel when command executed.
* Notification that command execution has begun.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1470,7 +1470,7 @@
                 {
                     "command": "aws.ecs.runCommandInContainer",
                     "group": "0@1",
-                    "when": "view == aws.explorer && viewItem == awsEcsContainerNode"
+                    "when": "view == aws.explorer && viewItem =~ /^(awsEcsContainerNodeExecEnabled|awsEcsContainerNodeExecDisabled)$/"
```
Comment (justinmk3) [inline package.json:1473]: they both are prefixed with `awsEcsContainerNodeExec`. could match only on that.

File: src/ecs/wizards/executeCommand.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import { createInputBox } from '../../shared/ui/inputPrompter'
+import { createQuickPick, DataQuickPickItem } from '../../shared/ui/pickerPrompter'
+import { localize } from '../../shared/utilities/vsCodeUtils'
+import { WIZARD_EXIT, Wizard, WIZARD_BACK } from '../../shared/wizards/wizard'
+import { EcsContainerNode } from '../explorer/ecsContainerNode'
+
+export interface CommandWizardState {
+    task: string
+    command: string
+    confirmation?: 'yes' | 'supress'
```
Comment (justinmk3) [inline src/ecs/wizards/executeCommand.ts:17]: misspelling (also in other parts of this PR)

```suggestion
    confirmation?: 'yes' | 'suppress'
```

File: src/ecs/wizards/executeCommand.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        // Filter for only 'Running' tasks
+        const runningTasks = (await node.describeTasks(taskArns)).filter(t => {
+            return t.lastStatus === 'RUNNING' && t.desiredStatus === 'RUNNING'
+        })
+        return runningTasks.map(task => {
+            // The last 32 digits of the task arn is the task identifier
+            return {
+                label: task.taskArn!.substr(-32),
+                detail: `Status: ${task.lastStatus}  Desired status: ${task.desiredStatus}`,
+                data: task.taskArn!.substr(-32),
```
Comment (justinmk3) [inline src/ecs/wizards/executeCommand.ts:35]: if `data` allows undefined, I would suggest `?` instead of `!`. If it doesn't allow undefined, coalesce to a defaultvalue such as empty string or a fixed string like "unknown".

Same for `label`.

```suggestion
                data: task.taskArn?.substr(-32) ?? 'unknown',
```

File: src/shared/utilities/childProcess.ts [hunk 4]
```diff
@@ -79,8 +79,8 @@ export class ChildProcess {
     private makeResult(code: number, signal?: string): ChildProcessResult {
         return {
             exitCode: code,
-            stdout: this.stdoutChunks.join().trim(),
-            stderr: this.stderrChunks.join().trim(),
+            stdout: this.stdoutChunks.join('').trim(),
+            stderr: this.stderrChunks.join('').trim(),
```
Comment (justinmk3) [inline src/shared/utilities/childProcess.ts:83]: this should be a separate commit. the other commits can be squashed into 1 commit

File: src/ecs/commands/runCommandInContainer.ts [hunk 5]
```diff
@@ -127,16 +48,34 @@ export async function runCommandInContainer(
...
         if (cp.exitCode !== 0) {
             result = 'Failed'
-            showOutputMessage(cp.stderr, outputChannel)
+            showOutputMessage(removeAnsi(cp.stderr), outputChannel)
             throw cp.error
         } else {
             result = 'Succeeded'
-            showOutputMessage(cp.stdout, outputChannel)
+            showOutputMessage(removeAnsi(cp.stdout), outputChannel)
         }
```
Comment (bryceitoc9) [inline src/ecs/commands/runCommandInContainer.ts:79]: Nit: is it worth showing both stderr and stdout no matter the situation? Error messages aren't just in stderr, and output isn't always limited to stdout.

File: src/ecs/commands/runCommandInContainer.ts [hunk 6]
```diff
@@ -127,16 +48,34 @@ export async function runCommandInContainer(
...
+            `${moment().format(LOCALIZED_DATE_FORMAT)}  Container: "${node.containerName}" Task ID: "${
+                response.task
+            }"  Running command: "${response.command}"`,
+            outputChannel
+        )
+
+        window.showInformationMessage(
+            localize(
+                'AWS.command.ecs.runCommandInContainer.runningCommandMessage',
+                'Running command. This may take several minutes.'
```
Comment (bryceitoc9) [inline src/ecs/commands/runCommandInContainer.ts:68]: Considerations:
* Add command name and container to message?
* "..." at the end instead of a period.
* Combined with feedback below, can we stream the output to the output channel and maybe provide a link here to open the output channel (as it is now, the output channel will open immediately, and switch back when the command completes...a button here would let the user swap away from the output channel and swap back)
  * If you do the above, consider outputting another notification on completion to signify the command is done with a link to the output channel.
* Consider converting to a progress (don't need to report specific progress increments, it's just nice to have the moving bar to show "progress" (also do indented bullet above if you swap to progress)

File: package.json (hunk 1 above)
Comment (bryceitoc9) [inline package.json:1473]: I believe this statement can be condensed to `/^(awsEcsContainerNodeExec)(.*)$/` . No worries though, we aren't super great about doing this throughout package.json anyway.

File: src/ecs/wizards/executeCommand.ts (hunk 2 above)
Comment (bryceitoc9) [reply src/ecs/wizards/executeCommand.ts:17]: Can this just be a boolean? `yes` and `suppress` are kind of strange as the two available values.

File: src/ecs/wizards/executeCommand.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+import { EcsContainerNode } from '../explorer/ecsContainerNode'
+
+export interface CommandWizardState {
+    task: string
+    command: string
+    confirmation?: 'yes' | 'supress'
+}
+
+function createTaskPrompter(node: EcsContainerNode) {
+    const taskItems = (async function () {
```
Comment (bryceitoc9) [inline src/ecs/wizards/executeCommand.ts:21]: Style nit: any reason why you're not using `async () => {` ?

File: src/ecs/wizards/executeCommand.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        const taskArns = await node.listTasks()
+        if (taskArns.length === 0) {
+            return []
+        }
+        // Filter for only 'Running' tasks
+        const runningTasks = (await node.describeTasks(taskArns)).filter(t => {
+            return t.lastStatus === 'RUNNING' && t.desiredStatus === 'RUNNING'
+        })
+        return runningTasks.map(task => {
```
Comment (bryceitoc9) [inline src/ecs/wizards/executeCommand.ts:30]: Nit: can be condensed to `return (await ....describeTasks.... ).filter(....filter block....).map(....map block....)`

File: package.json (hunk 1 above)
Comment (JadenSimon) [reply package.json:1473]: Once C9 adds `setContext` we can clean these clauses up.

File: src/ecs/wizards/executeCommand.ts (hunk 2 above)
Comment (JadenSimon) [reply src/ecs/wizards/executeCommand.ts:17]: Making this boolean would be more confusing as there are 3 options for this prompt. The third option ('no') is just 'hidden' from the type as it closes the wizard. 

For clarity we could make 'no' not close the wizard and instead handle it as its own thing. I think that could read better albeit with a bit more logic.

File: src/ecs/wizards/executeCommand.ts (hunk 3 above)
Comment (JadenSimon) [reply src/ecs/wizards/executeCommand.ts:35]: Even better would be to check for undefined closer to the source so these checks don't leak through the code.

File: src/ecs/wizards/executeCommand.ts (hunk 7 above)
Comment (JadenSimon) [reply src/ecs/wizards/executeCommand.ts:21]: In this case there's no functional difference, could go either way.

File: src/test/ecs/wizards/executeCommand.test.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+import * as _ from 'lodash'
+import { createWizardTester, WizardTester } from '../../../test/shared/wizards/wizardTestUtils'
+import { instance, mock } from '../../utilities/mockito'
+import { CommandWizard, CommandWizardState } from '../../../ecs/wizards/executeCommand'
+
+describe('CreateServiceWizard', function () {
+    let tester: WizardTester<CommandWizardState>
+    const node = mock()
```
Comment (JadenSimon) [inline src/test/ecs/wizards/executeCommand.test.ts:13]: Missing type. We should fix `mock` to use a class prototype to infer the type (separate PR)

File: src/ecs/wizards/executeCommand.ts (hunk 7 above)
Comment (bryceitoc9) [reply src/ecs/wizards/executeCommand.ts:21]: If there's no reason for `function ()`, bias towards consistent style.

File: package.json (hunk 1 above)
Comment (bryceitoc9) [reply package.json:1473]: Why?
* viewItem is supported and works fine in both IDEs
* There's no "right-click tree node" event hook so we can't evaluate context dynamically

Comment (JadenSimon) [reply package.json:1473]: I was talking more about some of the other clauses, not so much this one (though it's possible to use it for this one). Things like `copyArn` and `copyName` which can be handled in a cross-cutting way. As soon as we start composing contexts, `setContext` becomes desirable.

File: src/ecs/commands/runCommandInContainer.ts (hunk 6 above)
Comment (justinmk3) [reply src/ecs/commands/runCommandInContainer.ts:68]: >  Consider converting to a progress (don't need to report specific progress increments, it's just nice to have the moving bar to show "progress"

Yes. Separate PR though, please. This is in the backlog.

**Comment by justinmk3** [comment]
Looks good! Tested and it improves things.

> No notification that command is running

This is too noisy. Instead, just focus the output window, possibly also update the statusbar.

> * Datestamp and details added to output channel when command executed.

The localized datestamp should be the usual yyyy-MM-dd... format. `Jan 6, 2022 4:21:40 PM GMT-080` is hard to compare to other lines.
